### PR TITLE
Spotlight item

### PR DESCRIPTION
**Pull request by b0b3rt**
Create infrastructure for managing and displaying spotlight items.

image

Corresponding frontpage view:
image

Update one of the spotlight items to no longer be a draft:
image

Corresponding frontpage view:
image

## Engineering spec (mostly copied from slack)
MVP functionality desiderata:

- Can create a bunch of spotlight items ahead of time
- Spotlight items should have an "order" in which they're displayed
- We should switch to displaying the next item every X days (doesn't need to be configurable "per item" for the MVP)
- We should automatically cycle back to the start of the list if we run out
- We should be able to insert new items into arbitrary positions on the list
- We should be able to reorder existing items

Current design:

- have a position field (`number`) which indicates that item's ordering among all items
- have a lastPromotedAt field (`Date | null`) which indicates when an item was last "cycled in" to be displayed as the current spotlight item

To display on front page:
`SpotlightItems.findOne({}, { sort: { lastPromotedAt: -1 } })`

To cycle in a new item:

- Run a cron job (right now set to every 30 mins, but the period is almost irrelevant)
- Job runs the same query as the front page for determining the current spotlight item
- Then run a query to get the "next" one based on position/order
- Update the next one's lastPromotedAt to Date.now()

To add a new spotlight item in any position:

- If it's not given a position, query the max current position and add 1
- If it's given a position that's already occupied, shift everything at and after that position up by 1, then insert

Tests run:

- reposition the first spotlight item to be the last one
- reposition the last spotlight item to be the first one
- moving a spotlight item up by 1 position
- moving a spotlight item up by >1 position but not to be the first one
- moving a spotlight item down by 1 position
- moving a spotlight item down by >1 position by not to be the last one
- setting a spotlight item's position to be greater than the current max position +1
- having multiple non-drafted spotlight items, and checking that the one with the most recent `lastPromotedAt` is the one displayed on the frontpage
- having multiple non-drafted spotlight items, and checking that the one with the earliest position is the one displayed on the frontpage, if all the non-drafted spotlight items have the same `lastPromotedAt`

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203011033994129) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/form-components/ImageUpload.tsx [hunk 1]
```diff
@@ -80,6 +80,13 @@ const cloudinaryArgsByImageType = {
     minImageWidth: 480,
     cropping: false,
     uploadPreset: cloudinaryUploadPresetEventImageSetting.get()
+  },
+  spotlightImageId: {
+    minImageHeight: 232,
+    minImageWidth: 345,
```
Comment (b0b3rt) [inline packages/lesswrong/components/form-components/ImageUpload.tsx:86]: Double-check that these are the numbers we want

File: packages/lesswrong/components/form-components/ImageUpload.tsx [hunk 2]
```diff
@@ -80,6 +80,13 @@ const cloudinaryArgsByImageType = {
...
+    cropping: false,
+    // TODO: preset spotlight image url????
+    uploadPreset: cloudinaryUploadPresetGridImageSetting.get()
```
Comment (b0b3rt) [inline packages/lesswrong/components/form-components/ImageUpload.tsx:89]: Give this one its own db setting?  Not 100% sure what it's intended to be doing anyways

File: packages/lesswrong/components/recommendations/curatedContents.ts [hunk 3]
```diff
@@ -1,90 +1,90 @@
...
+// //   },
+// //   description: `<div><p>The Center for Applied Rationality set out to develop concepts and techniques that would (hopefully) result in clearer thinking and better decision-making.</p>
+// //   <p>This handbook is the result of the first five years or so of that research.</p></div>`,
+// //   imageUrl: "https://res.cloudinary.com/lesswrong-2-0/image/upload/v1660869601/DALL_E_2022-08-18_17.31.52_-_students_sitting_on_bean_bags_in_front_of_teacher_by_whiteboard_with_math_aquarelle_painting_by_JMW_Turner_-_edited2_zm0x5r.png",
+// //   firstPost: {
+// //     _id: "dbDHEQyKqnMDDqq2G",
+// //     url: "/s/KAv8z6oJCTxjR8vdR/p/dbDHEQyKqnMDDqq2G",
+// //     title: "Introduction"
+// //   }
+// // }
```
Comment (b0b3rt) [inline packages/lesswrong/components/recommendations/curatedContents.ts:90]: Delete the entire file, if it's unused?

File: packages/lesswrong/components/spotlights/SpotlightItem.tsx [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    },
+    [theme.breakpoints.down('sm')]: {
+      position: "absolute",
+      bottom: 4,
+      right: 8
+    },
+  },
+  editDescriptionButton: {
+
+  },
```
Comment (b0b3rt) [inline packages/lesswrong/components/spotlights/SpotlightItem.tsx:142]: Unused?

File: packages/lesswrong/components/spotlights/SpotlightItem.tsx [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                  fields={['description']}
+                  documentId={spotlight._id}
+                  mutationFragment={getFragment('SpotlightEditQueryFragment')}
+                  queryFragment={getFragment('SpotlightEditQueryFragment')}
+                  successCallback={() => setEdit(false)}
+                />
+              </div>
+              :
+              <div>
+                {/* <div className={classes.overflow}> */}
```
Comment (b0b3rt) [inline packages/lesswrong/components/spotlights/SpotlightItem.tsx:241]: cleanup?

File: packages/lesswrong/components/spotlights/SpotlightItem.tsx [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                  <ContentItemBody
+                    dangerouslySetInnerHTML={{__html: spotlight.description?.html ?? ''}}
+                    description={`${spotlight.documentType} ${spotlight.document._id}`}
+                  />
+                {/* </div> */}
```
Comment (b0b3rt) [inline packages/lesswrong/components/spotlights/SpotlightItem.tsx:246]: cleanup?

File: packages/lesswrong/components/spotlights/SpotlightsPage.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import Spotlights from '../../lib/collections/spotlights/collection';
+import { useMulti } from '../../lib/crud/withMulti';
+import { Components, getFragment, registerComponent } from '../../lib/vulcan-lib';
+
+const styles = (theme: ThemeType): JssStyles => ({
+  root: {
+
+  },
```
Comment (b0b3rt) [inline packages/lesswrong/components/spotlights/SpotlightsPage.tsx:9]: unused?

File: packages/lesswrong/server/spotlightCron.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      ? positionAscOrderedSpotlights.indexOf(lastSpotlightByPosition) + 1
+      : 0;
+
+    const { position: positionToPromote } = positionAscOrderedSpotlights[promoteIndex];
+
+    // This, of course, is not exactly the timestamp at which we actually promote this spotlight item.
+    // But doing it this way prevents a slow "walk" due to the cron job interval...
+    // ...ensuring the promotion happens at around the same time, and each period length is still roughly constant
+    const newPromotionDate = new Date(lastPromotionDate);
+    newPromotionDate.setDate(lastPromotionDate.getDate() + ACTIVE_DAYS);
```
Comment (jimrandomh) [inline packages/lesswrong/server/spotlightCron.ts:48]: Doing date-arithmetic this way sure makes me nervous, but I checked, and setting the day-of-the-month to >31 does wrap around to the next month correctly.

Comment (b0b3rt) [reply packages/lesswrong/server/spotlightCron.ts:48]: Yeah, I tested this and it seems to work correctly across both month (within-year) and month (across-year) boundaries.

**Review by Raemon — APPROVED** [review]

**Pull request merged**